src/governance-validators.cpp

```diff
@@ -0,0 +1,349 @@
+        return false;
+    }
+
+    CBitcoinAddress address(strPaymentAddress);
+    if(!address.IsValid()) {
+        strErrorMessages += "payment_address is invalid;";
+        return false;
+    }
+
+    return true;
+}
+
+bool CProposalValidator::ValidateURL()
+{
+    std::string strURL;
+    if(!GetDataValue("url", strURL)) {
+        strErrorMessages += "url field not found;";
+        return false;
+    }
+
+    std::string strURLStripped = StripWhitespace(strURL);
+
+    if(int(strURLStripped.size()) < 4) {
+        strErrorMessages += "url too short;";
+        return false;
+    }
+
+    if(!CheckURL(strURL)) {
+        strErrorMessages += "url invalid;";
+        return false;
+    }
+
+    return true;
+}
+
+void CProposalValidator::ParseJSONData()
+{
+    fJSONValid = false;
+
+    if(strData.empty()) {
+        return;
+    }
+
+    try {
+        UniValue obj(UniValue::VOBJ);
+        obj.read(strData);
+        std::vector<UniValue> arr1 = obj.getValues();
+        std::vector<UniValue> arr2 = arr1.at(0).getValues();
+        objJSON = arr2.at(1);
+        fJSONValid = true;
+    }
+    catch(std::exception& e) {
+        strErrorMessages += std::string(e.what()) + std::string(";");
+    }
+    catch(...) {
+        strErrorMessages += "Unknown exception;";
+    }
+}
+
+bool CProposalValidator::GetDataValue(const std::string& strKey, std::string& strValue)
+{
+    bool fOK = false;
+    try  {
+        strValue = objJSON[strKey].get_str();
+        fOK = true;
+    }
+    catch(std::exception& e) {
+        strErrorMessages += std::string(e.what()) + std::string(";");
+    }
+    catch(...) {
+        strErrorMessages += "Unknown exception;";
+    }
+    return fOK;
+}
+
+bool CProposalValidator::GetDataValue(const std::string& strKey, int64_t& nValue)
+{
+    bool fOK = false;
+    try  {
+        const UniValue uValue = objJSON[strKey];
+        switch(uValue.getType()) {
+        case UniValue::VNUM:
+            nValue = uValue.get_int64();
+            fOK = true;
+            break;
+        case UniValue::VSTR:
+        {
+            std::istringstream istr(uValue.get_str());
+            istr >> nValue;
+            fOK = ! istr.fail();
+        }
+        break;
+        default:
+            break;
+        }
+    }
+    catch(std::exception& e) {
+        strErrorMessages += std::string(e.what()) + std::string(";");
+    }
+    catch(...) {
+        strErrorMessages += "Unknown exception;";
+    }
+    return fOK;
+}
+
+bool CProposalValidator::GetDataValue(const std::string& strKey, double& dValue)
+{
+    bool fOK = false;
+    try  {
+        const UniValue uValue = objJSON[strKey];
+        switch(uValue.getType()) {
+        case UniValue::VNUM:
+            dValue = uValue.get_real();
+            fOK = true;
+            break;
+        case UniValue::VSTR:
+        {
+            std::istringstream istr(uValue.get_str());
+            istr >> dValue;
+            fOK = ! istr.fail();
```

Comment:
nitpick, inconsistent spacing.
Then again, ! does look like i and !i does my eyes in!
Feel free to ignore